cylc/flow/task_pool.py

```diff
@@ -585,116 +622,121 @@ def remove(self, itask, reason=""):
 
-    def get_task_by_id(self, id_):
-        """Return task with ID id_ if it exists, or None."""
-        for itask_ids in (
-                list(self.pool.values())
-                + list(self.runahead_pool.values())):
+    def _get_hidden_task_by_id(self, id_):
+        """Return runahead pool task by ID if it exists, or None."""
+        for itask_ids in list(self.hidden_pool.values()):
             try:
                 return itask_ids[id_]
             except KeyError:
                 pass
 
-    def queue_and_release(self):
-        self._queue_tasks()
-        return self._release_tasks()
+    def _get_task_by_id(self, id_):
+        """Return main pool task by ID if it exists, or None."""
+        for itask_ids in list(self.main_pool.values()):
+            try:
+                return itask_ids[id_]
+            except KeyError:
+                pass
 
-    def _queue_tasks(self):
+    def queue_tasks(self, itasks):
         """Queue tasks that are ready to run."""
-        queue_me = []
+        for itask in itasks:
+            itask.state.reset(is_queued=True)
+            # TODO Reset manual trigger flag. One manual trigger queues and
+            # unqueued task, another one triggers a queued task.
+            self.data_store_mgr.delta_task_state(itask)  # TODO needed?
+            self.data_store_mgr.delta_task_queued(itask)
+        self.task_queue_mgr.push_tasks(itasks)
+
+    def _queue_tasks(self):
+        """Queue tasks that are ready to run. (After reload?)"""
+        to_queue = []
         for itask in self.get_tasks():
             if itask.state.is_queued:
+                # Already queued
                 continue
-            ready_check_items = itask.is_ready()
+            ready_check_items = itask.is_ready_to_run()
+
+            # TODO: PUT THIS SOMEWHERE ELSE?:
             # Use this periodic checking point for data-store delta
             # creation, some items aren't event driven (i.e. clock).
             if itask.tdef.clocktrigger_offset is not None:
                 self.data_store_mgr.delta_task_clock_trigger(
                     itask, ready_check_items)
-            if all(ready_check_items):
-                queue_me.append(itask)
-                itask.state.reset(is_queued=True)
-                # Reset manual trigger flag. One manual trigger queues and
-                # unqueued task, another one triggers a queued task.
-                itask.reset_manual_trigger()
-                self.data_store_mgr.delta_task_state(itask)
-                self.data_store_mgr.delta_task_queued(itask)
 
-        self.task_queue_mgr.push_tasks(queue_me)
-        if queue_me:
-            LOG.debug(
-                "Queue pushed:\n"
-                + '\n'.join(f"* {t.identity}" for t in queue_me)
-            )
+            if all(ready_check_items) and not itask.state.is_runahead:
+                to_queue.append(itask)
 
-    def _release_tasks(self):
+        self.queue_tasks(to_queue)
```

Comment:
Why does the "private" `self._queue_tasks()` method call the "public" `self.queue_tasks()` and not the other way round?

Reply:
`_queue_tasks()` is only called by `reload_taskdefs()`. It figures out which tasks need to be queued, then calls the public `queue_tasks()` to do it.  I'll inline the private one into `reload_taskdefs()`, there's no good reason to have it as a separate method.

Reply:
(`_queue_tasks()` used to be called in the main loop, as well as at reload, to iterate the pool and find tasks to be queued. But now have event-driven queuing of tasks so that's not neded).